File: 200. Number of Islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    floodFill(grid, i, j);
                }
            }
        }
        return count;
    }
    private void floodFill(char[][] grid, int i, int j) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length) return;
        if (grid[i][j] == '0') return;
        grid[i][j] = '0';
        floodFill(grid, i+1, j);
        floodFill(grid, i-1, j);
        floodFill(grid, i, j+1);
        floodFill(grid, i, j-1);
    }
}

File: 200. Number of Islands.go
func numIslands(grid [][]byte) int {
    count := 0
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if (grid[i][j] == '1') {
                count++;
                floodFill(&grid, i, j)
            }
        }
    }
    return count
}

func floodFill(grid *[][]byte, i, j int) {
    if i < 0 || i >= len(*grid) {
        return
    }
    if j < 0 || j >= len((*grid)[0]) {
        return
    }
    if (*grid)[i][j] == '0' {
        return
    }
    (*grid)[i][j] = '0'
    floodFill(grid, i+1, j)
    floodFill(grid, i-1, j)
    floodFill(grid, i, j+1)
    floodFill(grid, i, j-1)

}
